### ISSUE TITLE
Homemade package + source control

### ISSUE THREAD

**Issue by erakis** [microsoft/vcpkg#2926]
Hi,

I read a lot of documentation about vcpkg and I two questions remain

1. I read the [zlib](https://github.com/Microsoft/vcpkg/blob/master/docs/examples/packaging-zlib.md) tutorial and I successfully created a **home made** package for zlib. But now, what If I want to keep this home made package available to other users (my work coworker) ? **What** do I have to keep in my source control (git) ? 

   I have some private package that I cannot push on github and should stay in our private git. So can I keep the my home made package recipes in git ? 

    But **how** and **where** ? I though pushing `vkpkg/port/private-lib-1` . So when coworkers would work on this project, they 
    1. Clone the project
    2. Clone the vcpkg repo at root of the project.

    This way, they got the `vcpkg/port/private-lib-1` already there. But is there a better way ? They I misunderstand ?

2. Pretend I want to get the library `rapidjson` but an old version. From what I've understand I need to modify the `vcpkg/ports/rapidjson/protfile.cmake` and change the url from there. But again, **same problem as 1)** how and what to backup to keep my recipe modification on my source control (git) ?

In the end, when a co-worker wants to work on the project, he should only
1. Clone the project
2. Read the ReadMe.md to understand that vcpkg is required and which library should be installed.
3. Install vcpkg
4. Install all vcpkg package listed from ReadMe.md.

Is this the right way to go?

**Comment by ras0219-msft** [comment]
Thanks for the questions!

The best option would be to upload your own vcpkg fork to your private git host. Then you can simply tell your coworker to pull that git fork and run bootstrap/install and you'll have 100% control over the versions of your libraries. Whenever you want updates, you can merge from the upstream repository here on GitHub and push them to your private fork. Until you pull from us again, you'll keep getting exactly the same libraries with exactly the same versions.

In this private fork, you can check in the portfiles (`ports\<libname>`) for your private libraries directly so they're easily available to your coworkers.

Additionally, if you'd like to share precompiled binaries instead, take a look at our `export` command[1].

[1] https://blogs.msdn.microsoft.com/vcblog/2017/05/03/vcpkg-introducing-export-command/

**Comment by purell** [comment]
Hi,

So if I well understand, I have to fork the `vcpkg` repo to my own git and then I add my custom package or modify existing one from there. Later I could pull news package from upstream (`vcpkg`) and merge existing one while keeping my own.

Did I understand properly? 
If *yes*, I think documentation should be updated with a section talking about custom package handling...

Best regards,

**Comment by ras0219-msft** [comment]
@purell We have some experimental support for Linux that we'd love to hear feedback on!

https://github.com/Microsoft/vcpkg/pull/1528#issuecomment-346975155

**Comment by ph55** [comment]
> Then you can simply tell your coworker to pull that git fork and run **bootstrap/install** and you'll have 100% control over the versions of your libraries

@ras0219-msft you mean `bootstrap-vcpkg.bat` file ? It also manages installed libraries ? Or you need to create your own file that install packages from list ?

**Comment by AugP** [comment]
Closing out old GitHub issue. If anyone is reading this today, we added some new features that can help those of you with custom port/library requirements: 
1. Registries feature: Allows you to point to a custom catalog of ports (e.g. your own private libraries, a forked copy of the vcpkg public catalog, etc.). See: https://vcpkg.io/en/docs/users/registries.html
2. Overlay ports: Allows you to use a custom variant of a port. Useful if you edited the source code. See: https://github.com/microsoft/vcpkg/blob/master/docs/specifications/ports-overlay.md
3. Versioning: Allows you to change the version of your ports (more convenient than doing it via overlay ports). See: https://vcpkg.io/en/docs/users/versioning.html
